include/zio/zio_fifo_buf.h

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) 2019 Thomas Burdick <thomas.burdick@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file zio_fifo_buf.h
+ *
+ * @brief ZIO FIFO Buffer provides a software zio_buf implementation
+ * based on a zio_fifo
+ */
+
+#ifndef ZEPHYR_INCLUDE_ZIO_FIFO_BUF_H_
+#define ZEPHYR_INCLUDE_ZIO_FIFO_BUF_H_
+
+#include <zio/zio_fifo.h>
+#include <zio/zio_buf.h>
+
+/**
+ * @private
+ * @brief Software FIFO implementation of a zio_buf
+ */
+struct z_zio_fifo_buf {
```

Comment:
I do not see a clear relation between this and zio_buf. For me this is one of the possible implementation of the zio_buf. So it should be accessed by zio_buf_XXX() API, but this is not the case. Also, the zio_buf defines some data structure which by my understanding is used by all implementations. However the k_poll_signal is defined here...

Reply:
Correct, this comes back to the indecision of where to place the pollable thing, I realized as I started creating this zio_buf implementation that it should probably be left up to the implementation itself rather than exposed in the zio_buf struct, but the indecision has been left in code here. I think it will become clearer as I start to implement a hardware fifo backed zio_buf for something like the icm 20649 I've been using. I'm pretty sure I want to leave it to the implementation rather than exposing it.

Reply:
I think this will be clearer when I update this PR with the polling changes you've suggested I make

Reply:
I think this should be clearer now given the additional test cases showing how polling might work. It needs more work as right now the semaphore works by showing the number of samples available, causing the semaphore to be pollable before the watermark is met. I'm thinking I might add polling modes in the future, but to start I'm going to just leave it as is and start working on example drivers